### DISCUSSION TITLE
`cue vet` playground ui

### DISCUSSION THREAD

**Discussion by disintegrator** [cue-lang/cue#1914] [Ideas]
It would be nice to have a [playground](https://cuelang.org/play/) UI for `cue vet` to support users that use CUE to write validations.

![image](https://user-images.githubusercontent.com/678548/189108260-acd5aaf3-8484-45e1-9f7a-cc82da9ee0ff.png)

**Comment by mvdan** [comment]
An alternative might be to lean into txtar files, similar to how Go's playground allows you to have multiple files: https://go.dev/play/p/BWJ4dcUqVfT

For example, one of the "Input" options could be "txtar", and then you could have something like:

```
cue vet foo.yaml foo.cue
-- foo.yaml --
[...]
-- foo.cue --
[...]
```

Then the output would be that of the command at the top. As a first rough pass, we could only allow one command at the top.

The reason I bring this up is because it's more flexible - it supports arbitrary input and output formats, any command that `cmd/cue` supports, etc. Then we don't need to keep expanding the dropdowns to support every possible combination a user might want to do.

I see that @myitcv already mentions txtar in a TODO in https://github.com/cue-lang/cuelang.org/tree/master/play; I'm not sure if this is what he had in mind.

**Reply by disintegrator** [reply]
Oh TIL. I've seen @myitcv use this format and I would personally be happy to use it. However, it's going to feel foreign to non-Go users and perhaps the UI can be a cover over the txtar representation...

**Reply by myitcv** [reply]
> However, it's going to feel foreign to non-Go users and perhaps the UI can be a cover over the txtar representation...

100% agree. For some people, they will prefer the light touch of `txtar`. For others it will be an implementation detail that can be hidden using good UI/UX. 

As you will have noticed from the playground, I score 0/10 on UI/UX. It was a weekend hack project early in the day when I was exploring CUE. Delighted to have others opine on how it can be improved.

**Comment by verdverm** [comment]
A devcontainer could be far less work with way more end user functionality

- https://docs.github.com/en/codespaces/setting-up-your-project-for-codespaces/introduction-to-dev-containers
- https://github.com/dagger/dagger/tree/main/.devcontainer

**Reply by myitcv** [reply]
Definitely something we should look at. Indeed, as @disintegrator mentioned on Slack, I don't think investigating this route is mutually exclusive with improving the playground.
